### PR TITLE
Fix NullableMemberTranslator to look only at Nullable<T>

### LINKED ISSUE #18486: Query: Incorrect match for Value?
Following repro: https://github.com/aspnet/EntityFrameworkCore/issues/14180#issuecomment-544193017
Following seems to be matching `.Value` to our nullable.Value and ignoring that Value is a property which we cannot translate.
```C#
context.Entities.SingleOrDefault(e => e.Name == name && e.Version.Value == version)
```

### PR DESCRIPTION
**Pull request by roji**
NullableMemberTranslator was using IsNullableType with the intent to check for Nullable<T>, but that method returns true also for reference types. Since I've also ran into this confusion myself in the past, I've renamed IsNullableType (internal) to IsNullableOrReferenceType to better reflect what it does.

Fixes #18486

### CODE REVIEW

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 1]
```diff
@@ -7063,6 +7063,61 @@ public BugContext13517(DbContextOptions options)
 
         #endregion
 
+        #region Issue14180
+
+        [Fact]
+        public void Value_conversion_with_property_named_value()
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs]: CustomConvertersTestBase

Comment (roji) [reply test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs]: Done

**Comment by roji** [comment]
@smitpatel after being moved the test now fails on InMemory (it's cranky today...) - there's no exception being thrown. Am not familiar enough with this, is it expected behavior for InMemory to allow accessing unknown properties etc (and do we intend for it to stay that way)? If so I'll skip the test for InMemory.

**Comment by smitpatel** [comment]
@roji - https://github.com/aspnet/EntityFrameworkCore/issues/17050

**Comment by roji** [comment]
Thanks @smitpatel

File: src/Shared/SharedTypeExtensions.cs [hunk 2]
```diff
@@ -16,19 +16,16 @@ internal static class SharedTypeExtensions
         public static Type UnwrapNullableType(this Type type) => Nullable.GetUnderlyingType(type) ?? type;
 
         public static bool IsNullableType(this Type type)
```
Comment (smitpatel) [inline src/Shared/SharedTypeExtensions.cs]: Names I would prefer
`IsNullableType` - any type where I can assign null.
`IsNullableValueType`/`IsNullableReferenceType` (in future)

Comment (roji) [reply src/Shared/SharedTypeExtensions.cs]: No problem, done the change.

**Pull request merged**